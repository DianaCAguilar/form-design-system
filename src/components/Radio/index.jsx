import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import cc from 'classcat';

import RadioEmptyIcon from 'lib/icons/react/RadioEmptyIcon';
import RadioFilledIcon from 'lib/icons/react/RadioFilledIcon';

const Radio = React.forwardRef(
  (
    {
      showLabel = true,
      disabled = false,
      name,
      label,
      checked,
      value,
      onChange,
      className,
      ...rest
    },
    ref
  ) => {
    const RadioIcon = useMemo(() => {
      if (checked) return RadioFilledIcon;
      return RadioEmptyIcon;
    }, [checked]);

    return (
      <div
        className={cc([
          {
            'fdsCheckable--disabled': disabled,
            'fdsRadio--checked': checked,
          },
          'fdsCheckable fdsRadio',
        ])}
      >
        <label className="flush--bottom">
          <input
            ref={ref}
            type="radio"
            name={name}
            checked={checked}
            value={value}
            onChange={onChange}
            className="media--xs"
            {...rest}
          />
          <span className="fdsCheckable-icon--checked">
            <RadioIcon size="xs" />
          </span>
          {showLabel && <span className="padding--left--s">{label}</span>}
        </label>
      </div>
    );
  }
);

Radio.displayName = 'Radio';

Radio.propTypes = {
  /**
   * html `name` attribute.
   * Use to create radio groups.
   */
  name: PropTypes.string.isRequired,

  /** html `label` element content */
  label: PropTypes.string.isRequired,

  /** If the supplied `label` prop should be rendered to the screen. */
  showLabel: PropTypes.bool,

  /** Disables form field when `true` */
  disabled: PropTypes.bool,

  /** The radio input `checked` attribute */
  checked: PropTypes.bool,

  /** The radio input `value` attribute */
  value: PropTypes.string.isRequired,

  /** The radio input `onChange` attribute */
  onChange: PropTypes.func,

  className: PropTypes.string,
};

export default Radio;
