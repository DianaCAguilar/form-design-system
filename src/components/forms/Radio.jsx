import React, { useState } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import RadioEmptyIcon from '../../../lib/icons/react/RadioEmptyIcon';
import RadioFilledIcon from '../../../lib/icons/react/RadioFilledIcon';

/**
 * @param {Object} props react props
 * @returns {ReactElement}
 */
const Radio = ({
  name,
  label,
  showLabel,
  value,
  onChange,
  checked,
  disabled,
  ...otherProps
}) => {
  const [isChecked, setIsChecked] = useState(checked);

  const handleChange = () => {
    if (!disabled) {
      const updatedCheckedState = !isChecked;
      setIsChecked(updatedCheckedState);
      onChange(updatedCheckedState);
    }
  };

  const Icon = isChecked ? RadioFilledIcon : RadioEmptyIcon;

  return (
    <div className={cx('fdsCheckable fdsRadio', { 'fdsCheckable--disabled': disabled })}>
      <input
        type="radio"
        name={name}
        id={name}
        value={value}
        className="media--xs"
        onChange={handleChange}
        checked={isChecked}
        disabled={disabled}
        {...otherProps}
      />
      <label className="flush--bottom" htmlFor={name}>
        <div role="radio" aria-checked={isChecked} aria-label={label}>
          <Icon size="xs" />
        </div>
        {showLabel && <span className="padding--left--half">{label}</span>}
      </label>
    </div>
  );
};

Radio.defaultProps = {
  value: undefined,
  checked: false,
  disabled: false,
  showLabel: true,
  onChange: () => {},
};

Radio.propTypes = {
  /**
   * `name` attribute of input element.
   * Also used for the `htmlFor` prop of the `label`.
   */
  name: PropTypes.string.isRequired,

  /** Label used for a11y attributes _and_ the rendered `label` element */
  label: PropTypes.string.isRequired,

  /** If the supplied `label` prop should be rendered to the screen. */
  showLabel: PropTypes.bool,

  /** optional accessibility label - will not render any text */
  a11yLabel: PropTypes.string,

  /** optional `value` attribute of input element */
  value: PropTypes.string,

  /** onChange callback - invoked with the checked state of the radio:
   * ```
   * <Radio onChange={(isChecked) => {}} ... />
   * ```
   */
  onChange: PropTypes.func,

  /** Sets `checked` state */
  checked: PropTypes.bool,

  /** Disables form field when `true` */
  disabled: PropTypes.bool,
};

export default Radio;
