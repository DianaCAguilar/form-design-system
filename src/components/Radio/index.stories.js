import React, { useState } from 'react';
import PropTypes from 'prop-types';
import Radio from '.';

const Template = (args) => <Radio {...args} value="1" />;
const MultiTemplate = ({ multi = {}, ...args }) => {
  const [isChecked, setIsChecked] = useState('1');
  const onChange = (e) => {
    setIsChecked(e.target.value);
  };
  return (
    <div onChange={onChange}>
      <Radio {...args} {...multi.one} checked={isChecked === '1'} value="1" />
      <Radio {...args} {...multi.two} checked={isChecked === '2'} value="2" />
      <Radio {...args} {...multi.three} checked={isChecked === '3'} value="3" />
    </div>
  );
};

MultiTemplate.propTypes = {
  multi: PropTypes.object,
};

export const Primary = Template.bind({});
Primary.args = {
  name: 'radio-group-name',
  label: 'I agree to receive spam',
};

export const RadioGroups = MultiTemplate.bind({});
RadioGroups.args = {
  name: 'radio-group-story',
  multi: {
    one: { label: 'Option one' },
    two: { label: 'Option two' },
    three: { label: 'Option three' },
  },
};

RadioGroups.parameters = {
  docs: {
    description: {
      story:
        'You can create radio groups by rendering `Radio` components that share the same `name` prop.',
    },
  },
};

export default {
  component: Radio,
  title: 'components/Radio',
  argTypes: { onChange: { action: 'onChange' } },
  parameters: {
    docs: {
      fds: {
        attributes: ['rest'],
      },
      description: {
        component: 'Uncontrolled custom radio component with label.',
      },
    },
  },
};
