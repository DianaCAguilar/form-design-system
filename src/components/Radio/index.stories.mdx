import React from 'react';

import Radio from '.';

const Template = args => <Radio {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  name: "radio-group-name",
  label: "I agree to receive spam"
}

export const radioGroups = args => (
  <React.Fragment>
    <Radio {...args} name="radio-group-story" label="Option one" />
    <Radio {...args} name="radio-group-story" label="Option two" />
    <Radio {...args} name="radio-group-story" label="Option three" />
  </React.Fragment>
)

radioGroups.parameters = {
  docs: {
    description: {
      story:
        'You can create radio groups by rendering `Radio` components that share the same `name` prop.',
    },
  },
};

export const initialSelection = args => (
  <React.Fragment>
    <Radio {...args} name="radio-group-initial" label="Selected by default" defaultChecked />
    <Radio {...args} name="radio-group-initial" label="The other option" />
  </React.Fragment>
)

initialSelection.parameters = {
  docs: {
    description: {
      story:
        'Use the `defaultChecked` prop to set initial selection in a radio group.',
    },
  },
};

export const readingSelectedValue = args => (
  <div onChange={(e) => { console.warn(`value is ${e.target.value}`) }}>
    <Radio {...args} name="radio-group-onchange" label="Option one" value="1" />
    <Radio {...args} name="radio-group-onchange" label="Option two" value="2" />
    <Radio {...args} name="radio-group-onchange" label="Option three" value="3" />
  </div>
)

readingSelectedValue.parameters = {
  docs: {
    description: {
      story:
        'Place the `onChange` event on a parent element to read which radio of the group has been selected.'
    },
  },
};

export const controlled = args => (
    <div onChange={(e) => { console.warn(`value is ${e.target.value}`) }}>
      <Radio {...args} name="radio-group-onchange" label="Option one" value="1" checked={true} />
      <Radio {...args} name="radio-group-onchange" label="Option two" value="2" />
      <Radio {...args} name="radio-group-onchange" label="Option three" value="3" />
    </div>
)

export default {
  component: Radio,
  title: 'components/Radio',
  parameters: {
    docs: {
      description: {
        component: 'Uncontrolled custom radio component with label.'
      },
    },
  },
};


