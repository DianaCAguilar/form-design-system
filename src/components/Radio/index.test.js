import React from 'react';
import { shallow } from 'enzyme';

import Radio from '.';

describe('Radio component', () => {
  it('matches snapshot', () => {
    const wrapper = shallow(
      <Radio label="Snapshot Label" name="snapshot-radio" value="test-value" />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('matches snapshot (disabled)', () => {
    const wrapper = shallow(
      <Radio disabled label="Snapshot Label" name="snapshot-radio" value="test-value" />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('matches snapshot (checked)', () => {
    const wrapper = shallow(
      <Radio label="Snapshot Label" name="snapshot-radio" value="test-value" />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('fires change callback when checking', () => {
    const changeFn = jest.fn();
    const input = shallow(
      <Radio
        name="callback-test"
        onChange={changeFn}
        label="Test change"
        value="test-value"
      />
    ).find('input');

    expect(changeFn).not.toHaveBeenCalled();
    input.simulate('change', { target: { checked: true } });
    expect(changeFn).toHaveBeenCalled();
  });
});
