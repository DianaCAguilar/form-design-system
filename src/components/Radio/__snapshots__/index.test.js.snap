// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio component matches snapshot (checked) 1`] = `
<div
  className="fdsCheckable fdsRadio"
>
  <label
    className="flush--bottom"
  >
    <input
      className="media--xs"
      name="snapshot-radio"
      type="radio"
      value="test-value"
    />
    <span
      className="fdsCheckable-icon--checked"
    >
      <RadioEmptyIcon
        size="xs"
      />
    </span>
    <span
      className="padding--left--s"
    >
      Snapshot Label
    </span>
  </label>
</div>
`;

exports[`Radio component matches snapshot (disabled) 1`] = `
<div
  className="fdsCheckable--disabled fdsCheckable fdsRadio"
>
  <label
    className="flush--bottom"
  >
    <input
      className="media--xs"
      name="snapshot-radio"
      type="radio"
      value="test-value"
    />
    <span
      className="fdsCheckable-icon--checked"
    >
      <RadioEmptyIcon
        size="xs"
      />
    </span>
    <span
      className="padding--left--s"
    >
      Snapshot Label
    </span>
  </label>
</div>
`;

exports[`Radio component matches snapshot 1`] = `
<div
  className="fdsCheckable fdsRadio"
>
  <label
    className="flush--bottom"
  >
    <input
      className="media--xs"
      name="snapshot-radio"
      type="radio"
      value="test-value"
    />
    <span
      className="fdsCheckable-icon--checked"
    >
      <RadioEmptyIcon
        size="xs"
      />
    </span>
    <span
      className="padding--left--s"
    >
      Snapshot Label
    </span>
  </label>
</div>
`;
